``Project 1`` :
--

Analyse the dataset of "Census Income" of year 1994, by solving many queries.

( Data set source : https://archive.ics.uci.edu/ml/datasets/Adult )

**``Limited Googling`` is allowed. Max time limit for this project 60 mins**  

**Adult_DataSet has following features :**

``age`` : continuous.

``workclass`` : Private, Self-emp-not-inc, Self-emp-inc, Federal-gov, Local-gov, State-gov, Without-pay, Never-worked.

``fnlwgt`` : continuous.

``education`` : Bachelors, Some-college, 11th, HS-grad, Prof-school, Assoc-acdm, Assoc-voc, 9th, 7th-8th, 12th, Masters, 1st-4th, 10th, Doctorate, 5th-6th, Preschool.

``education-num`` : continuous.

``marital-status`` : Married-civ-spouse, Divorced, Never-married, Separated, Widowed, Married-spouse-absent, Married-AF-spouse.

``occupation`` : Tech-support, Craft-repair, Other-service, Sales, Exec-managerial, Prof-specialty, Handlers-cleaners, Machine-op-inspct, Adm-clerical, Farming-fishing, Transport-moving, Priv-house-serv, Protective-serv, Armed-Forces.

``relationship`` : Wife, Own-child, Husband, Not-in-family, Other-relative, Unmarried.

``race`` : White, Asian-Pac-Islander, Amer-Indian-Eskimo, Other, Black.

``sex`` : Female, Male.

``capital-gain`` : continuous.

``capital-loss`` : continuous.

``hours-per-week`` : continuous.

``native-country`` : United-States, Cambodia, England, Puerto-Rico, Canada, Germany, Outlying-US(Guam-USVI-etc), India, Japan, Greece, South, China, Cuba, Iran, Honduras, Philippines, Italy, Poland, Jamaica, Vietnam, Mexico, Portugal, Ireland, France, Dominican-Republic, Laos, Ecuador, Taiwan, Haiti, Columbia, Hungary, Guatemala, Nicaragua, Scotland, Thailand, Yugoslavia, El-Salvador, Trinadad&Tobago, Peru, Hong, Holand-Netherlands.

``salary`` : >50K,<=50K

https://www.quora.com/What-is-the-meaning-of-capital-gain-capital-loss-and-fnlwgt-in-adult-dataset-from-UCI

In [2]:
import pandas as pd

In [3]:
data = pd.read_csv('C:\Program Files/Python36/suven/Data Anaylsis/Dataset/adult_data.csv')
data.head()

age          workclass  fnlwgt   education  education-num  \
0   39          State-gov   77516   Bachelors             13   
1   50   Self-emp-not-inc   83311   Bachelors             13   
2   38            Private  215646     HS-grad              9   
3   53            Private  234721        11th              7   
4   28            Private  338409   Bachelors             13   

        marital-status          occupation    relationship    race      sex  \
0        Never-married        Adm-clerical   Not-in-family   White     Male   
1   Married-civ-spouse     Exec-managerial         Husband   White     Male   
2             Divorced   Handlers-cleaners   Not-in-family   White     Male   
3   Married-civ-spouse   Handlers-cleaners         Husband   Black     Male   
4   Married-civ-spouse      Prof-specialty            Wife   Black   Female   

   capital-gain  capital-loss  hours-per-week  native-country  salary  
0          2174             0              40   United-States   <=50K  
1             0             0              13   United-States   <=50K  
2             0             0              40   United-States   <=50K  
3             0             0              40   United-States   <=50K  
4             0             0              40            Cuba   <=50K

In [4]:
# 1> How many men and women are represented in this dataset?
# Hint : use value_counts()

data['sex'].value_counts()


 Male      21790
 Female    10771
Name: sex, dtype: int64

In [5]:
# 2. What is the average age (age feature) of women?
# Note :  gender is stored as " Male"  and " Female". Watch is the space ahead.

data.loc[data['sex']==' Female','age'].mean()

36.85823043357163

In [6]:
# 3. What is the proportion of German citizens (native-country feature)?
# Note : Specify country as ' Germany'. Note the space ahead.

# Hint : 
# total no. of entries in the "Income Census"
# data.shape[0]

# proportion of Germans = Count_of_Germans / total_population
float((data['native-country']==' Germany').sum())/data.shape[0]

0.004207487485028101

In [7]:
# 4. What are mean value and standard deviation of the age of those 
#    who recieve more than 50K per year (salary feature) 
#    and those who receive less than 50K per year?

ages1 = data.loc[data['salary'] == ' >50K', 'age']
ages2 = data.loc[data['salary'] == ' <=50K', 'age']

print("The average age of the rich: {0} +- {1} years, poor - {2} +- {3} years.".format(
    round(ages1.mean(), 2), round(ages1.std(), 2),
    round(ages2.mean(), 2), round(ages2.std(), 2)))

The average age of the rich: 44.25 +- 10.52 years, poor - 36.78 +- 14.02 years.


In [11]:
# 5. Is it true that people who receive more than 50k have 
# at least high school education? 
# (education - Bachelors, Prof-school, Assoc-acdm, Assoc-voc, 
# Masters or Doctorate feature)

data.loc[data['salary'] == ' >50K', 'education'].unique()

# Observation or ans :  

array([' HS-grad', ' Masters', ' Bachelors', ' Some-college',
       ' Assoc-voc', ' Doctorate', ' Prof-school', ' Assoc-acdm',
       ' 7th-8th', ' 12th', ' 10th', ' 11th', ' 9th', ' 5th-6th',
       ' 1st-4th'], dtype=object)

In [12]:
# 6. Display statistics of age for each race (race feature) and each gender. 
# Use groupby() and describe(). 
# Find the maximum age of men of Amer-Indian-Eskimo race.

for (race,sex), sub_df in data.groupby(['race','sex']):
    print("Race:{0}, sex: {1}".format(race, sex))
    print(sub_df['age'].describe())

    
    
# ans : maximum age of men of Amer-Indian-Eskimo race =     

Race: Amer-Indian-Eskimo, sex:  Female
count    119.000000
mean      37.117647
std       13.114991
min       17.000000
25%       27.000000
50%       36.000000
75%       46.000000
max       80.000000
Name: age, dtype: float64
Race: Amer-Indian-Eskimo, sex:  Male
count    192.000000
mean      37.208333
std       12.049563
min       17.000000
25%       28.000000
50%       35.000000
75%       45.000000
max       82.000000
Name: age, dtype: float64
Race: Asian-Pac-Islander, sex:  Female
count    346.000000
mean      35.089595
std       12.300845
min       17.000000
25%       25.000000
50%       33.000000
75%       43.750000
max       75.000000
Name: age, dtype: float64
Race: Asian-Pac-Islander, sex:  Male
count    693.000000
mean      39.073593
std       12.883944
min       18.000000
25%       29.000000
50%       37.000000
75%       46.000000
max       90.000000
Name: age, dtype: float64
Race: Black, sex:  Female
count    1555.000000
mean       37.854019
std        12.637197
min        17.0

In [13]:
# 7. <Solved>
# Among whom the proportion of those who earn a lot(>50K) is more: 
# among married or single men (marital-status feature)? 
# Consider married those who have a marital-status starting with 
# Married (Married-civ-spouse, Married-spouse-absent or Married-AF-spouse), 
# the rest are considered bachelors.

data.loc[(data['sex'] == ' Male') &
     (data['marital-status'].isin([' Never-married', 
                                   ' Separated', 
                                   ' Divorced',
                                   ' Widowed'])), 'salary'].value_counts()

 <=50K    7552
 >50K      697
Name: salary, dtype: int64

In [15]:
data.loc[(data['sex'] == ' Male') &
     (data['marital-status'].str.startswith(' Married')), 'salary'].value_counts()

 <=50K    7576
 >50K     5965
Name: salary, dtype: int64

So their are like **5965 "MARRIED"** people of high income group (>50k) vs. just **697 UnMARRIED** people of similar high income group. 

``Ans`` : More rich people are Married people (& hopefully Happy too !!)

In [16]:
# just for overall stats -> how many of each category ?
data['marital-status'].value_counts()

 Married-civ-spouse       14976
 Never-married            10683
 Divorced                  4443
 Separated                 1025
 Widowed                    993
 Married-spouse-absent      418
 Married-AF-spouse           23
Name: marital-status, dtype: int64

In [18]:
# <solved>
# 8. What is the maximum number of hours a person works per week 
# (hours-per-week feature)? 
# How many people work such a number of hours and 
# what is the percentage of those who earn a lot among them?

# finding maximum hours-per-week
max_load = data['hours-per-week'].max()
print("Max time - {0} hours./week.".format(max_load))


# finding those working max_load hours-per-week. Lets call it num_workaholics  
num_workaholics = data[data['hours-per-week'] == max_load].shape[0]
print("Total number of such hard workers {0}".format(num_workaholics))


# finding percentage of those who earn a lot among them 
rich_share = float(data[(data['hours-per-week'] == max_load)
                 & (data['salary'] == '>50K')].shape[0]) / num_workaholics

num=float(data[(data['hours-per-week'] == max_load)
                 & (data['salary'] == '>50K')].shape[0])

print(num)
print("Percentage of rich among them {0}%".format(float(100 * rich_share)))

Max time - 99 hours./week.
Total number of such hard workers 85
0.0
Percentage of rich among them 0.0%


In [20]:
# 9. Count the average time of work (hours-per-week) those 
# who earning a little and a lot (salary) for each country (native-country) ?

# Hint 1 :  use    .groupby(['native-country', 'salary']
# Hint 2 :  then perform  temp_dataframe['hours-per-week'].mean()

for (country, salary), sub_df in data.groupby(['native-country', 'salary']):
    print(country, salary, round(sub_df['hours-per-week'].mean(),2))



 ?  <=50K 40.16
 ?  >50K 45.55
 Cambodia  <=50K 41.42
 Cambodia  >50K 40.0
 Canada  <=50K 37.91
 Canada  >50K 45.64
 China  <=50K 37.38
 China  >50K 38.9
 Columbia  <=50K 38.68
 Columbia  >50K 50.0
 Cuba  <=50K 37.99
 Cuba  >50K 42.44
 Dominican-Republic  <=50K 42.34
 Dominican-Republic  >50K 47.0
 Ecuador  <=50K 38.04
 Ecuador  >50K 48.75
 El-Salvador  <=50K 36.03
 El-Salvador  >50K 45.0
 England  <=50K 40.48
 England  >50K 44.53
 France  <=50K 41.06
 France  >50K 50.75
 Germany  <=50K 39.14
 Germany  >50K 44.98
 Greece  <=50K 41.81
 Greece  >50K 50.62
 Guatemala  <=50K 39.36
 Guatemala  >50K 36.67
 Haiti  <=50K 36.33
 Haiti  >50K 42.75
 Holand-Netherlands  <=50K 40.0
 Honduras  <=50K 34.33
 Honduras  >50K 60.0
 Hong  <=50K 39.14
 Hong  >50K 45.0
 Hungary  <=50K 31.3
 Hungary  >50K 50.0
 India  <=50K 38.23
 India  >50K 46.48
 Iran  <=50K 41.44
 Iran  >50K 47.5
 Ireland  <=50K 40.95
 Ireland  >50K 48.0
 Italy  <=50K 39.62
 Italy  >50K 45.4
 Jamaica  <=50K 38.24
 Jamaica  >50K 41.1
 Jap

**From the above data, we get country, salary wise , mean Hours-per-week**

 ---------------------------- **This ends your first project work** ----------------------------